_posts/2020-04-18-regression-chains/regression-chains.Rmd

```diff
@@ -0,0 +1,179 @@
+
+```{r}
+dat = as.data.table(cbind(Y, x1, x2))
+str(dat)
+```
+
+## 3D Visualization of the Data
+
+If you feel confident to already have a good feeling of the data, feel free to skip this section. If not, you can use the `rgl` package to play around with the following three 3D plots with the feature variables on the x- and y-axis and the target variables on the respective z-axes:
+
+```{r, eval = FALSE}
+library(rgl)
+colfun = colorRampPalette(c("#161B1D", "#ADD8E6"))
+```
+
+```{r, eval = FALSE}
+setorder(dat, y1)
+plot3d(dat$x1, dat$x2, dat$y1,
+  xlab = "x1", ylab = "x2", zlab = "y1",
+  type = "s", radius = 0.1, col = colfun(n))
+```
+
+```{r, eval = FALSE}
+setorder(dat, y2)
+plot3d(dat$x1, dat$x2, dat$y2,
+  xlab = "x1", ylab = "x2", zlab = "y2",
+  type = "s", radius = 0.1, col = colfun(n))
+```
+
+```{r, eval = FALSE}
+setorder(dat, y3)
+plot3d(dat$x1, dat$x2, dat$y3,
+  xlab = "x1", ylab = "x2", zlab = "y3",
+  type = "s", radius = 0.1, col = colfun(n))
+```
+
+## Building the Pipeline
+
+In our regression chain, the first model will predict $y_{1}$. Therefore, we initialize our task with respect to this target:
+
+```{r}
+library(mlr3)
+library(mlr3learners)
+library(mlr3pipelines)
+task = TaskRegr$new("multiregression", backend = dat, target = "y1")
+```
+
+As learners we will use simple linear regression models, `r ref("mlr_learners_regr.lm", "lm learner")`. Our pipeline building the regression chain then has to do the following:
+
+* Use the input to predict $y_{1}$ within the first learner (i.e., $y_{1} \sim x_{1} + x_{2} + y_{2} + y_{3}$).
+* Combine the input with the prediction of $y_{1}$, $\hat{y_{1}}$ and use this to predict $y_{2}$ within the second learner (i.e., $y_{2} \sim x_{1} + x_{2} + \hat{y_{1}} + y_{3}$).
+* Combine the input with the prediction of $y_{2}$ and use this to predict $y_{3}$ within the final learner (i.e., $y_{3} \sim x_{1} + x_{2} + \hat{y_{1}} + \hat{y_{2}}$).
+
+To combine predictions of a learner with the previous input, we rely on `PipeOpLearnerCV` and `PipeOpNOP` arranged in parallel via `gunion` combined via `PipeOpFeatureUnion`, e.g., in the first step:
+
+```{r}
+step1 = po("copy", outnum = 2, id = "copy1") %>>%
+  gunion(list(
+    po("learner_cv", learner = lrn("regr.lm"), id = "y1_learner"),
+    po("nop", id = "nop1")
+  )) %>>%
+  po("featureunion", id = "union1")
+step1$plot()
+```
+
+Within the second step we then have to define $y_{2}$ as the new target. This can be done using `PipeOpUpdateTarget` (note that `PipeOpUpdateTarget` by default drops the original target, here $y_{1}$). Typically one would use `PipeOpUpdateTarget` to apply a transformation function to the original target (possibly also renaming it) which must not be inverted later. However, in the special case of leaving the `trafo` at its default `identity` and providing a `new_target_name` that already matches a column of the input data, `PipeOpUpdateTarget` simply sets this column as the new target:
+
+```{r}
+#FIXME:
```

Comment:
Perhaps for now, just delete it and say in a comment that this is currently not exported, but will be in a future version.